### ISSUE TITLE
Autogenerated ShadowMesh uses same LoD level as source Mesh

### ISSUE THREAD

**Issue by smix8** [godotengine/godot#47859]
**Godot version:**
Godot 4.x
commit b4b7c97d3839170180e4f3f7c2bd3179b374057a

**OS/device including version:**
Win 10 64x

**Issue description:**
After testing the new ShadowMeshes in Godot 4.x and comparing the new feature with manual ShadowMeshes created for Godot 3.x I couldn't help but feel very disappointed by the performance (~80% less than Godot 3.x).

So I looked at the shadowmesh source code and I consider this an implementation bug and not a missing feature for a proposal.

Autogenerated shadowmeshes in Godot 4.x are matched to the same source mesh LoD level while they should be matched 1-3 LoD levels lower. A shadow imposter mesh created manual by an artist always starts at least 1-2 LoD levels below the visible LoD mesh but never at the same level. A typical example would be a characters hair mesh, starting at 150.000 vertex. The shadow looks the same with a LOD3 mesh that has only 5.000 vertex cause the outside mesh shape stays the same and this is enough for shadow casting on solid objects. In the current implementation the shadow casting mesh has a 145.000 vertex overhead that does nothing for the endresult and just consumes performance.

Due to this as mentioned in #45452 pull request there is currently only a little performance gain with shadow imposters while it should be a considerable one.

I suggest that the autogenerated shadow imposters should be generated and matched with LOD2 or LOD3 meshes by default to fix this issue and give the expected performance boost from shadow imposters.

There is no reason to ever use the same mesh geometry detail for a shadow-only casting object except for edgecases like meshes mixed with semi-transparent surfaces and very lowpoly meshes (that don't need shadow imposters to begin with). A Shadowmesh LoD selection option in the new ResourceImporter would be welcome to solve those edgecases.

**Steps to reproduce:**

**Minimal reproduction project:**

**Comment by Calinou** [comment]
cc @fire

**Comment by mrjustaguy** [comment]
Eh, #45452 isn't about generating Shadow Mesh LODs...

as @reduz said:
> -When importing, a vertex-only version of the mesh is created.
> -This version is used when rendering shadows, and improves performance by reducing bandwidth

This doesn't Mean that the Generated Mesh is supposed to be a LOD mesh. If I understand what the Optimization is doing (Not a Rendering Expert) It's Giving The Renderer Only Verts to render the shadow, instead of triangles, because with triangles it wouldn't be uncommon for a vertex to be a part of multiple triangles, resulting in multiple copies of the same vertex, which increases bandwidth usage, and thus reduces performance.

**Comment by clayjohn** [comment]
@mrjustaguy is very close, ShadowMeshes are not the same thing as ShadowImposters. ShadowMeshes are a copy of the mesh that contain vertex positions and indices only (rather than including normals, tangents, UVs, etc.). ShadowMeshes contain all the same LODs as the regular mesh. 

I think there needs to be a proposal for adding a "Shadow LOD bias" parameter somewhere that shifts which LOD to use for the shadow meshes. Perhaps an import setting?

**Comment by mrjustaguy** [comment]
Could be a Mesh Instance Setting, like LOD bias is.

**Comment by smix8** [comment]
@mrjustaguy Yep you are right, the linked pullrequest is not directly the same but I thought it is involved.
@clayjohn thanks for the more detailed information.

I think one issue is that the terms ShadowMesh and ShadowImposter are mixed all the time by people that have no technical render background. I see experiences 3D artists use both terms all the time when they mean ShadowImposters and the user interface in Godot 4.x uses the term "Generate -> Shadow Meshes" right next to LOD generation.

I think a pulldown option in the new Advanced Import Settings menu with all available mesh LOD levels could work to pick the starting mesh detail for the shadow meshes / imposters.
Something like "ShadowMesh Base LOD Level".

Set to default LOD1 or LOD2 if available on the mesh.

![shadowmeshes](https://user-images.githubusercontent.com/52464204/114579913-a13eec80-9c7e-11eb-9f00-97b989f496bd.JPG)

**Comment by mrjustaguy** [comment]
Currently a Workaround would be (if Cast Shadow property, under geometry, worked, which it doesn't seem to do in current G4.0) is to make 2 mesh instances, a Mesh that is visible but doesn't cast shadows, and a shadow mesh that is set to just cast shadows, with a lower LOD Bias.
